### PR TITLE
Malonso/durable objects

### PR DESCRIPTION
**Pull request by xortive**
- [x] use class and script name for DO bindings instead of namespace id
- [x] adhoc create/delete/unused migration support
- [ ] tagged create/delete/unused migration support
    - [ ] fetch current migration tag from API
    - [ ] coalesce migrations from wrangler.toml into single API migration
- [ ] support for rename/transfer migrations

### CODE REVIEW

File: src/commands/publish.rs [hunk 1]
```diff
@@ -112,6 +88,7 @@ pub fn publish(
 
         let upload_client = http::featured_legacy_auth_client(user, Feature::Sites);
 
+        deploy::pre_upload()?;
```
Comment (ags799) [inline src/commands/publish.rs:91]: Why are we putting this empty function here and elsewhere?

**Review by ags799 — APPROVED** [review]

**Review by caass — COMMENTED** [review]
Looks pretty good! A couple things we can chat about on monday but overall happy with the way this looks -- is there a PR in the works to update docs as well?

File: src/commands/publish.rs (hunk 1 above)
Comment (xortive) [reply src/commands/publish.rs:91]: `pre_upload()` is for doing tasks that require making an API call before uploading the script. This is going to be used to fetch the current migration tag for tagged migrations, so wrangler can look at the list of migrations in `wrangler.toml` and know how to coalesce them.

I had already done some of the work for tagged migrations and this is part of it, so I figured i'd leave it in

File: src/settings/toml/durable_objects.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
+use serde::{Deserialize, Serialize};
+
+use crate::settings::binding::Binding;
+
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct DurableObjects {
+    pub classes: Option<Vec<DurableObjectsClass>>,
+    #[serde(skip_deserializing)] // TODO remove when tagged migrations are fully supported
```
[the remark this answers is not in the document]
Comment (xortive) [reply src/settings/toml/durable_objects.rs:8]: migrations are how you tell the platform to
* activate (for lack of a better word) a durable objects class such that you can address it, and create durable objects implemented by that class
* delete all of the durable objects implemented by a durable objects class
* rename durable objects classes
* transfer durable objects between classes

right now wrangler only supports untagged migrations specified when publishing -- this is fine for small projects, but teams will need tagged migrations that can be committed to version control.

The migration tag just serves as a way to get the latest migration that the platform has actually been executed, so when you `publish` with several un-executed migrations after the latest one in `wrangler.toml`, those migrations can get combined into a single migration.

**Comment by xortive** [comment]
Pre-merge tasks:

- [ ] help text for migration directives
- [ ] change modules support to only treat .bin as an arraybuffer -- unknown extensions are an error
- [ ] rename deploy output helper to run_deploy

**Pull request merged**